doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,80 @@
+Format:
+
+```
+    %% FUNCTION-BASE-NAME
+
+    prologue
+
+    % PARAM-NAME type [flags]
+    code to process param
+
+    % PARAM-NAME type [flags]
+    code to process param
+
+    %
+
+    epilogue
+```
+
+- the % characters must be start of line
+
+- FUNCTION-BASE-NAME is the name the functions will derive from e.g. "kdf_scrypt" to this will be appended _[gs]et[_ctx]_params
+
+- type is the standard OSSL_PARAM types with additions for basic C types
+
+- Flags include:
+  - alias (uses same name as another entry but different type)
+  - multiple (multiple instances are allowed - not sure if necessary)
+
+
+The idea is that the gettable and get functions will be simultaneously generated
+along with a perfect hash to decode the legal parameters quickly.
+
+
+
+A trivial example:
+```
+    generate_ctx_get(p{
+        %% kdf_scrypt get ctx
+        # no prologue
+        %% KDF_PARAM_SIZE size_t
+            if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_SIZE)) != NULL)
+                return OSSL_PARAM_set_size_t(p, SIZE_MAX);
+        %%
+        # no epilogue
+    });
+```
+
+Would generate something along the lines of:
+```c
+    enum { kdf_scrypt_ctx_get_param_KDF_PARAM_SIZE };
+
+    static const int kdf_scrypt_ctx_get_param_lookup(const OSSL_PARAM params[],
+                                                     int *indicies) {
+        ...
+    }
+
+    static const OSSL_PARAM *kdf_scrypt_gettable_ctx_params(ossl_unused void *ctx,
+                                                            ossl_unused void *p_ctx)
+    {
+        static const OSSL_PARAM known_gettable_ctx_params[] = {
+            OSSL_PARAM_size_t(OSSL_KDF_PARAM_SIZE, NULL),
+            OSSL_PARAM_END
+        };
+        return known_gettable_ctx_params;
+    }
+
+    static int kdf_scrypt_get_ctx_params(void *vctx, OSSL_PARAM params[])
+    {
+        OSSL_PARAM *p;
+        int indicies[1];
```

Comment:
Yeah. I was thinking the codegen could '#define' a 'NUM_INDICES' or similar for the handwritten code to use to declare the array on the stack (or alternatively an array typedef), allowing the right length for each implementation. But if the entire function is being generated by codegen that's not needed. :+1: